src/calliope/preprocess/model_math.py

```diff
@@ -58,79 +59,116 @@ def __iter__(self):
-            raise exceptions.ModelError(
-                f"Attempted to override math with pre-defined math file '{filename}'."
+        Raises:
+            ModelError: If math has already been applied.
+        """
+        if self.in_history(filename):
+            raise ModelError(
+                f"Math preprocessing | Overwriting with previously applied pre-defined math: '{filename}'."
             )
-        self._add_file(MATH_DIR / f"{filename}.yaml", filename)
+        with importlib.resources.as_file(
+            importlib.resources.files("calliope") / "math"
+        ) as f:
+            self._add_file(f / f"{filename}.yaml", filename)
 
-    def add_user_defined_file(
+    def _add_user_defined_file(
         self, relative_filepath: str | Path, model_def_path: str | Path
     ) -> None:
-        """Add user-defined Calliope math, relative to the model definition path."""
+        """Add user-defined Calliope math, relative to the model definition path.
+
+        Args:
+            relative_filepath (str | Path): Path to user math, relative to model definition.
+            model_def_path (str | Path): Model definition path.
+
+        Raises:
+            ModelError: If file has already been applied.
+        """
         math_name = str(relative_filepath)
-        if self.check_in_history(math_name):
-            raise exceptions.ModelError(
-                f"Attempted to override math with user-defined math file '{math_name}'"
+        if self.in_history(math_name):
+            raise ModelError(
+                f"Math preprocessing | Overwriting with previously applied user-defined math: '{relative_filepath}'."
             )
         self._add_file(relative_path(model_def_path, relative_filepath), math_name)
 
-    def validate(self, extra_math: dict | None = None):
-        """Test current math and optional external math against the MATH schema.
+    def _init_from_string(
+        self, math_to_add: str, model_def_path: str | Path | None = None
+    ):
+        """Load math definition from a list of files.
 
         Args:
-            extra_math (dict | None, optional): Temporary math to merge into the check. Defaults to None.
+            math_to_add (str): Calliope math file to load. Suffix implies user-math.
+            model_def_path (str | Path | None, optional): Model definition path. Defaults to None.
+
+        Raises:
+            ModelError: User-math requested without providing `model_def_path`.
         """
-        math_to_validate = deepcopy(self.data)
-        if extra_math is not None:
-            math_to_validate.union(AttrDict(extra_math), allow_override=True)
-        validate_dict(math_to_validate, MATH_SCHEMA, "math")
-        LOGGER.info("ModelMath: validated math against schema.")
+        if not math_to_add.endswith((".yaml", ".yml")):
+            self._add_pre_defined_file(math_to_add)
+        else:
+            self._add_user_defined_file(math_to_add, model_def_path)
```

Comment:
Bug: does not catch invalid call when requesting user defined math without a def_path.
```suggestion
        if not math_to_add.endswith((".yaml", ".yml")):
            self._add_pre_defined_file(math_to_add)
        elif model_def_path:
            self._add_user_defined_file(math_to_add, model_def_path)
        else:
            raise ModelError("Must define model definition path when requesting user math.")
```